File: dodam-core/src/main/java/b1nd/dodamcore/nightstudy/application/NightStudyService.java
package b1nd.dodamcore.nightstudy.application;

import b1nd.dodamcore.member.domain.entity.Student;
import b1nd.dodamcore.nightstudy.domain.entity.NightStudy;
import b1nd.dodamcore.nightstudy.domain.enums.NightStudyStatus;
import b1nd.dodamcore.nightstudy.domain.exception.NightStudyNotFoundException;
import b1nd.dodamcore.nightstudy.repository.NightStudyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NightStudyService {

    private final NightStudyRepository repository;

    public void save(NightStudy nightStudy) {
        repository.save(nightStudy);
    }

    public void delete(NightStudy nightStudy) {
        repository.delete(nightStudy);
    }

    public NightStudy getBy(Long id) {
        return repository.findById(id)
                .orElseThrow(NightStudyNotFoundException::new);
    }

    public boolean checkDurationDuplication(Student student, LocalDate startAt, LocalDate endAt) {
        return repository.existsValidByStudentAndDate(student, startAt, endAt);
    }

    public List<NightStudy> getMy(Student student, LocalDate now) {
        return repository.findByStudentAndEndAtGreaterThanEqual(student, now);
    }

    public List<NightStudy> getPending() {
        return repository.findByStatus(NightStudyStatus.PENDING);
    }

    public List<NightStudy> getValid(LocalDate now) {
        return repository.findAllowedStudyByDate(now, NightStudyStatus.ALLOWED);
    }

}

File: services/service-night-study/src/main/kotlin/com/b1nd/dodamdodam/nightstudy/domain/nightstudy/service/NightStudyService.kt
package com.b1nd.dodamdodam.nightstudy.domain.nightstudy.service

import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.entity.NightStudyEntity
import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.entity.NightStudyMemberEntity
import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.enumeration.NightStudyStatus
import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.enumeration.NightStudyType
import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.exception.NightStudyNotFoundException
import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.repository.NightStudyMemberRepository
import com.b1nd.dodamdodam.nightstudy.domain.nightstudy.repository.NightStudyRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.UUID

@Service
class NightStudyService(
    private val nightStudyRepository: NightStudyRepository,
    private val nightStudyMemberRepository: NightStudyMemberRepository
) {

    companion object {
        val INDIVIDUAL_TYPES = listOf(NightStudyType.NIGHT_STUDY_1, NightStudyType.NIGHT_STUDY_2, NightStudyType.NIGHT_STUDY_3)
        val PROJECT_TYPES = listOf(NightStudyType.PROJECT_1, NightStudyType.PROJECT_2)
    }

    fun save(nightStudy: NightStudyEntity): NightStudyEntity =
        nightStudyRepository.save(nightStudy)

    fun getById(id: Long): NightStudyEntity =
        nightStudyRepository.findById(id).orElseThrow { NightStudyNotFoundException() }

    fun softDelete(nightStudy: NightStudyEntity) {
        nightStudy.softDelete()
        nightStudyRepository.save(nightStudy)
    }

    // Individual night study queries
    fun getByUserId(userId: UUID): List<NightStudyEntity> =
        nightStudyRepository.findAllByUserIdAndTypeIn(userId, INDIVIDUAL_TYPES)

    fun getAllowed(date: LocalDate): List<NightStudyEntity> =
        nightStudyRepository.findAllByStatusAndTypeInAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
            NightStudyStatus.ALLOWED, INDIVIDUAL_TYPES, date, date
        )

    fun getPending(): List<NightStudyEntity> =
        nightStudyRepository.findAllByStatusAndTypeIn(NightStudyStatus.PENDING, INDIVIDUAL_TYPES)

    fun getAll(): List<NightStudyEntity> =
        nightStudyRepository.findAllByTypeIn(INDIVIDUAL_TYPES)

    fun getProjectsByUserId(userId: UUID): List<NightStudyEntity> =
        nightStudyRepository.findAllByUserIdAndTypeIn(userId, PROJECT_TYPES)

    fun getAllowedProjects(date: LocalDate): List<NightStudyEntity> =
        nightStudyRepository.findAllByStatusAndTypeInAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
            NightStudyStatus.ALLOWED, PROJECT_TYPES, date, date
        )

    fun getPendingProjects(): List<NightStudyEntity> =
        nightStudyRepository.findAllByStatusAndTypeIn(NightStudyStatus.PENDING, PROJECT_TYPES)

    // Member management
    fun saveMembers(members: List<NightStudyMemberEntity>): List<NightStudyMemberEntity> =
        nightStudyMemberRepository.saveAll(members)

    fun getMembersByNightStudyId(nightStudyId: Long): List<NightStudyMemberEntity> =
        nightStudyMemberRepository.findAllByNightStudyId(nightStudyId)

    fun getMembersByNightStudyIds(nightStudyIds: List<Long>): List<NightStudyMemberEntity> =
        nightStudyMemberRepository.findAllByNightStudyIdIn(nightStudyIds)
}
